include/boost/multiprecision/detail/fpclassify.hpp

```diff
@@ -0,0 +1,69 @@
+///////////////////////////////////////////////////////////////////////////////
+//  Copyright 2022 Matt Borland. Distributed under the Boost
+//  Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MP_DETAIL_FPCLASSIFY_HPP
+#define BOOST_MP_DETAIL_FPCLASSIFY_HPP
+
+#include <cmath>
+#include <limits>
+#include <boost/multiprecision/detail/standalone_config.hpp>
+#include <boost/multiprecision/detail/number_base.hpp>
+
+#ifdef BOOST_MP_MATH_AVAILABLE
+#include <boost/math/special_functions/fpclassify.hpp>
+
+namespace boost { namespace multiprecision { namespace detail {
+
+template <typename T>
+inline bool isnan BOOST_PREVENT_MACRO_SUBSTITUTION (const T val)
+{
+    return (boost::math::isnan)(val);
+}
+
+template <typename T>
+inline bool isinf BOOST_PREVENT_MACRO_SUBSTITUTION (const T val)
+{
+    return (boost::math::isinf)(val);
+}
+
+template <typename T>
+inline int fpclassify BOOST_PREVENT_MACRO_SUBSTITUTION (const T val)
+{
+    return (boost::math::fpclassify)(val);
+}
+
+}}} // Namespaces
+
+#else
+
+namespace boost { namespace multiprecision { namespace detail {
+
+// Assume IEEE 754 compliance and compiler optimizations do not affect this
+template <typename T>
+inline constexpr bool isnan BOOST_PREVENT_MACRO_SUBSTITUTION (const T val) noexcept
+{
+    return val != val;
```

Comment:
I'm hoping we can ditch this whole header, but the val != val trick is known not to work for some optimisation levels.  Actually std::isnan breaks when some optimisations are enabled for gcc as well, but we have no control over that :(